src/providers/virtual/qgsvirtuallayerprovider.cpp

```diff
@@ -317,7 +317,9 @@ bool QgsVirtualLayerProvider::createIt()
       provider.replace( QLatin1String( "'" ), QLatin1String( "''" ) );
       QString source = mLayers.at( i ).source;
       source.replace( QLatin1String( "'" ), QLatin1String( "''" ) );
-      QString encoding = mLayers.at( i ).encoding;
+      QString encoding = mLayers.at( i ).encoding.isEmpty()
+                         ? QStringLiteral( "System" )
+                         : mLayers.at( i ).encoding;
```

Comment:
When the virtual layer is broken, the encoding becomes an empty string, which causes a broken SQL below in the `QgsVLayer('%2','%4',%3)` part, as `%3` is nothing and the trailing comma is wrong (and dangerous).